<a href="https://colab.research.google.com/github/tinkvu/Data-Acquisition-Dublin.ie/blob/main/DublinIE.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Acquisition of Dublin.ie

## Introduction

Welcome to the **Dublin.ie** Data Acquisition Project!
This initiative is dedicated to capturing the heartbeat of Dublin through its vibrant and diverse events. The focus of this project is to extract information about upcoming events in Dublin from the "https://dublin.ie/whats-on/" page. By leveraging web scraping and Beautiful Soup within a Jupyter Notebook on Google Colab, I aim to provide a comprehensive overview of the exciting events that contribute to Dublin's dynamic cultural scene.


**Project Objective**

Dublin.ie serves as the primary source for our data acquisition project, specifically targeting the "What's On" section. This is where Dublin's upcoming events are showcased, ranging from cultural festivals and entertainment to educational seminars.
This project collects datas from the website as from the date of running and makes the data in a convenient way for all to do different types of processing and analysis.

**The Essence of Dublin Events**

Dublin is a city that thrives on its cultural richness, and events play a pivotal role in shaping its character. Each event contributes to the diverse and dynamic tapestry of Dublin's cultural landscape. This project aspires to capture the essence of Dublin through the lens of its upcoming events, providing a glimpse into the vibrant and multifaceted activities that define the city.

**How It Works**

The project remains organized into sections within a Jupyter Notebook, with a dedicated focus on extracting and processing event information. Users are encouraged to explore the notebook to uncover details about specific events, venues, dates, and other relevant information. The project's success lies in its ability to provide a real-time snapshot of the events that make Dublin an exciting and lively city.

**Join the Exploration**

Whether you are planning your visit to Dublin, a local seeking new experiences, or a researcher interested in Dublin's cultural scene, this project invites you to join the exploration of upcoming events. As we delve into the data, let's uncover the richness of Dublin's events and celebrate the city's cultural diversity.

Let the event exploration begin!

##Data Acquisition Codes

Let's start grabbing datas.

First we import the requests function to have a connection to the website.

We will be using BeautifulSoup package for parsing the html to the python code

Before getting started, we have to go to the target website and check what kind of queries they have used to pass data and by checking that only, we can get to a conclusion for the methods we use.

In [ ]:
import requests
from bs4 import BeautifulSoup
import pandas as pd

url = "https://dublin.ie/whats-on/"  #URL that is gonna scrape
resp = requests.get(url)
soup=BeautifulSoup(resp.text,'html.parser') #Here the 'soup' retrieves the Html of the website
soup

In [ ]:
events=[] #events is our list in which we are going to store the datas collected

#The below 2 lines holds the important part of the website where the required datas are stored
events_container = soup.find('div', class_='articles events-listing autofit')
all_events = events_container.find_all('article', class_='event card')

#This for loop traverses through all the datas we need to acquire and stores it
for event in all_events:

    event_link = event.find('a', class_='overlay')['href']
    image_url = event.find('img')['src']
    event_name = event.find('h2').text
    location = event.find('p', class_='location').text.strip()
    date_range = event.find('time').text.strip()
    summary = event.find('div', class_='summary').find('p').text.strip()
    category = event['data-categories']

    #Appending the datas in the list we created
    events.append({
    'Event Name': event_name,
    'Category' : category,
    'Location': location,
    'Date Range': date_range,
    'Description': summary,
    'Image URL': image_url,
    'Event Link': event_link

})

#Lets make the list into a dataframe
events=pd.DataFrame(events)
print(events.shape)
events.head()

(114, 7)


Event Name  \
0                             Describe the Night   
1                               Really Good Time   
2  Gifted – The Contemporary Craft & Design Fair   
3             Sustainable Christmas Craft Market   
4            A Festive Evening with Sing Ireland   

                                            Category  \
0                                    art-and-theatre   
1                                   music-and-comedy   
2  art-and-theatre,christmas,family-friendly,food...   
3                                          christmas   
4         christmas,family-friendly,music-and-comedy   

                           Location                   Date Range  \
0                Glass Mask Theatre   Mon 20th Nov - Sat 9th Dec   
1                   The Sound House                  Sat 9th Dec   
2                               RDS   Wed 6th Dec - Sun 10th Dec   
3          National Botanic Gardens   Sat 9th Dec - Sun 10th Dec   
4  EPIC The Irish Emigration Museum  Sun 19th Nov - Sun 10th Dec   

                                         Description  \
0  A theatrical odyssey spanning 90 years. A desp...   
1  Really Good Time is a four-piece, apres-garde,...   
2  The Iconic Christmas Craft Fair is Back, Featu...   
3  The traditional Christmas Craft Market returns...   
4  This Christmas, EPIC Museum are glad to partne...   

                                           Image URL  \
0  https://cdn.dublin.ie/wp-content/uploads/WO-Gl...   
1  https://cdn.dublin.ie/wp-content/uploads/WO-So...   
2  https://cdn.dublin.ie/wp-content/uploads/WO-Gi...   
3  https://cdn.dublin.ie/wp-content/uploads/WO-Bo...   
4  https://cdn.dublin.ie/wp-content/uploads/WO-EP...   

                                          Event Link  
0  https://dublin.ie/whats-on/listings/describe-t...  
1  https://dublin.ie/whats-on/listings/really-goo...  
2  https://dublin.ie/whats-on/listings/gifted-the...  
3  https://dublin.ie/whats-on/listings/christmas-...  
4  https://dublin.ie/whats-on/listings/a-festive-...

Wuhoooo!

Now we have got 114 event datas from the website with all the major informations like Name, Category, Location, Dates and so on!

**Data Preprocessing:**

The datas we got needs to be processed since it needs to be in a better analytical forms. So here starts the preprocessing steps.

Let us make the Date Range into a DateTime Object

In [ ]:
from dateutil import parser

# Function to clean and standardize the date format
def clean_date(date_range):
    date_list = date_range.split('-')
    sDate = parser.parse(date_list[0].strip())
    eDate = parser.parse(date_list[-1].strip()) if len(date_list) > 1 else sDate
    return pd.Series({'sDate': sDate, 'eDate': eDate})

# Apply the function to the 'Date Range' column
events = pd.concat([events, events['Date Range'].apply(clean_date)], axis=1)
events.drop('Date Range',axis=1, inplace=True)
#print(type(events['Date Range']))
df=events

#Now we have dropped the Date Range column and made it as 2 columns, sDate and eDate
df[["sDate", "eDate"]] = df[["sDate", "eDate"]].apply(pd.to_datetime)
df.head()

Event Name  \
0                             Describe the Night   
1                               Really Good Time   
2  Gifted – The Contemporary Craft & Design Fair   
3             Sustainable Christmas Craft Market   
4            A Festive Evening with Sing Ireland   

                                            Category  \
0                                    art-and-theatre   
1                                   music-and-comedy   
2  art-and-theatre,christmas,family-friendly,food...   
3                                          christmas   
4         christmas,family-friendly,music-and-comedy   

                           Location  \
0                Glass Mask Theatre   
1                   The Sound House   
2                               RDS   
3          National Botanic Gardens   
4  EPIC The Irish Emigration Museum   

                                         Description  \
0  A theatrical odyssey spanning 90 years. A desp...   
1  Really Good Time is a four-piece, apres-garde,...   
2  The Iconic Christmas Craft Fair is Back, Featu...   
3  The traditional Christmas Craft Market returns...   
4  This Christmas, EPIC Museum are glad to partne...   

                                           Image URL  \
0  https://cdn.dublin.ie/wp-content/uploads/WO-Gl...   
1  https://cdn.dublin.ie/wp-content/uploads/WO-So...   
2  https://cdn.dublin.ie/wp-content/uploads/WO-Gi...   
3  https://cdn.dublin.ie/wp-content/uploads/WO-Bo...   
4  https://cdn.dublin.ie/wp-content/uploads/WO-EP...   

                                          Event Link      sDate      eDate  
0  https://dublin.ie/whats-on/listings/describe-t... 2023-11-20 2023-12-09  
1  https://dublin.ie/whats-on/listings/really-goo... 2023-12-09 2023-12-09  
2  https://dublin.ie/whats-on/listings/gifted-the... 2023-12-06 2023-12-10  
3  https://dublin.ie/whats-on/listings/christmas-... 2023-12-09 2023-12-10  
4  https://dublin.ie/whats-on/listings/a-festive-... 2023-11-19 2023-12-10

Now, let us categorize the datas in the **Category column**. There are different categories the Category column and in some of the column, there are more than 1 categories. So, making it into a dictionary will help us in Analysis.

In [ ]:
def convert_to_dict(category):
    categories = category.split(',')
    return {'category': categories}

# Apply the function to the 'Category' column
df['Category'] = df['Category'].apply(convert_to_dict)

In [ ]:
df.head()

Event Name  \
0                             Describe the Night   
1                               Really Good Time   
2  Gifted – The Contemporary Craft & Design Fair   
3             Sustainable Christmas Craft Market   
4            A Festive Evening with Sing Ireland   

                                            Category  \
0                  {'category': ['art-and-theatre']}   
1                 {'category': ['music-and-comedy']}   
2  {'category': ['art-and-theatre', 'christmas', ...   
3                        {'category': ['christmas']}   
4  {'category': ['christmas', 'family-friendly', ...   

                           Location  \
0                Glass Mask Theatre   
1                   The Sound House   
2                               RDS   
3          National Botanic Gardens   
4  EPIC The Irish Emigration Museum   

                                         Description  \
0  A theatrical odyssey spanning 90 years. A desp...   
1  Really Good Time is a four-piece, apres-garde,...   
2  The Iconic Christmas Craft Fair is Back, Featu...   
3  The traditional Christmas Craft Market returns...   
4  This Christmas, EPIC Museum are glad to partne...   

                                           Image URL  \
0  https://cdn.dublin.ie/wp-content/uploads/WO-Gl...   
1  https://cdn.dublin.ie/wp-content/uploads/WO-So...   
2  https://cdn.dublin.ie/wp-content/uploads/WO-Gi...   
3  https://cdn.dublin.ie/wp-content/uploads/WO-Bo...   
4  https://cdn.dublin.ie/wp-content/uploads/WO-EP...   

                                          Event Link      sDate      eDate  
0  https://dublin.ie/whats-on/listings/describe-t... 2023-11-20 2023-12-09  
1  https://dublin.ie/whats-on/listings/really-goo... 2023-12-09 2023-12-09  
2  https://dublin.ie/whats-on/listings/gifted-the... 2023-12-06 2023-12-10  
3  https://dublin.ie/whats-on/listings/christmas-... 2023-12-09 2023-12-10  
4  https://dublin.ie/whats-on/listings/a-festive-... 2023-11-19 2023-12-10

Now let us export the file to a Comma Seperated Values (csv)

In [ ]:
df.to_csv('Dublin.ie Events.csv')

## Filtering events by date

Now let's do some filtering.

What if I wanna know about the events between 2 specific dates?

In [ ]:
start_date = pd.to_datetime('2023-12-01') #You can add any date here in the format YYY-MM-DD
end_date = pd.to_datetime('2023-12-10')  # Here the end date
filtDf = df[(df['sDate'] >= start_date) & (df['eDate'] <= end_date)]
filtDf

#Now by running this line, we will get all the events in this specific dates

Event Name  \
1                                    Really Good Time   
2       Gifted – The Contemporary Craft & Design Fair   
3                  Sustainable Christmas Craft Market   
5                        Stillgarden Christmas Market   
7                       The Fumbally Christmas Market   
8          IAFF – Ibero-American Film Festival Dublin   
9   The Irish National Youth Ballet perform Cinder...   
37                                    Sleeping Beauty   
40                   Cinderella- The Civic Panto 2023   
41    Charlie and the Chocolate Factory – The Musical   
44       SO Fine Art Editions: Winter Exhibition 2023   
52                   ‘Medici Lion’ by Siobhán Hapaska   
53  MELLOW TONICS Christmas Concert: Pure Imagination   
54               Ardgillan Castle Christmas Gift Fair   
59                     Sammy, Buffy and the Beanstalk   
77                       Olly, Polly & the Magic Lamp   
81                                         Funderland   
83                                 The Sound of Music   
84                                         NYF Dublin   

                                             Category  \
1                  {'category': ['music-and-comedy']}   
2   {'category': ['art-and-theatre', 'christmas', ...   
3                         {'category': ['christmas']}   
5   {'category': ['business-and-tech', 'christmas'...   
7    {'category': ['business-and-tech', 'christmas']}   
8   {'category': ['festivals', 'film-and-literatur...   
9      {'category': ['art-and-theatre', 'christmas']}   
37  {'category': ['art-and-theatre', 'christmas', ...   
40  {'category': ['art-and-theatre', 'christmas', ...   
41  {'category': ['christmas', 'family-friendly', ...   
44                  {'category': ['art-and-theatre']}   
52   {'category': ['art-and-theatre', 'free-events']}   
53    {'category': ['christmas', 'music-and-comedy']}   
54  {'category': ['christmas', 'family-friendly', ...   
59  {'category': ['art-and-theatre', 'christmas', ...   
77  {'category': ['art-and-theatre', 'christmas', ...   
81     {'category': ['christmas', 'family-friendly']}   
83  {'category': ['art-and-theatre', 'christmas', ...   
84  {'category': ['art-and-theatre', 'christmas', ...   

                    Location  \
1            The Sound House   
2                        RDS   
3   National Botanic Gardens   
5     Stillgarden Distillery   
7               The Fumbally   
8                 UCD Cinema   
9     Samuel Beckett Theatre   
37              Mill Theatre   
40             Civic Theatre   
41  Bord Gáis Energy Theatre   
44      SO Fine Art Editions   
52  The Douglas Hyde Gallery   
53    Pepper Canister Church   
54          Ardgillan Castle   
59      The National Stadium   
77          3Olympia Theatre   
81                       RDS   
83     National Concert Hall   
84         Various Locations   

                                          Description  \
1   Really Good Time is a four-piece, apres-garde,...   
2   The Iconic Christmas Craft Fair is Back, Featu...   
3   The traditional Christmas Craft Market returns...   
5   A Christmas Market in the heart of Dublin 8 sh...   
7   This year's market is a phenomenal cross secti...   
8   From December 6 to 10, 2023, Dublin is gearing...   
9   The Irish National Youth Ballet perform Cinder...   
37  “Sleep my Beauty, that’s what I decree. No pre...   
40  The Fairy Godmother Of All Pantos!\nThe team w...   
41  Escape to a world of pure imagination with ROA...   
44  The annual Winter Exhibition 2023 showcases so...   
52  The Douglas Hyde is delighted to present a maj...   
53  Step into the whimsical and enchanting world o...   
54  Soak up the ultimate Christmas Atmosphere in A...   
59  Celebrating is 25TH Anniversary THIS YEAR. Joi...   
77  It’s that time of year again! “Oh Yes it is”!\...   
81  The One and Only Funderland returns to where i...   
83  After a sensational sell-out run last year, th...   
84  Dublin City Will Come Alive In Sp

## Filtering by categories

Now what if I want to filter this by a specific category?

Let's see what are the different categories in the column:

In [ ]:
categories_list = df['Category'].apply(lambda x: x.get('category', []))
categories_flat = [category for sublist in categories_list for category in sublist]
unique_categories = set(categories_flat)
unique_categories

{'art-and-theatre',
 'business-and-tech',
 'christmas',
 'family-friendly',
 'festivals',
 'film-and-literature',
 'food-and-drink',
 'free-events',
 'learning',
 'music-and-comedy',
 'online',
 'sports'}

There are 12 different categories in the data.

Now comes the question: Filter out the **free events**.

In [ ]:
desired_categories = ['free-events']
filtDf = df[df['Category'].apply(lambda x: any(category in x['category'] for category in desired_categories))]
filtDf

Event Name  \
5               Stillgarden Christmas Market   
6                       The Enchanted Garden   
15                    Christmas at Farmleigh   
16           Bremore Castle Christmas Market   
19                   Christmas at the Castle   
30     Live Animal Crib at the Mansion House   
32  Dublin City Council Dublin Winter Lights   
38                           The Moving Crib   
51   Solidarity – The Dockers of Dublin Port   
52          ‘Medici Lion’ by Siobhán Hapaska   
54      Ardgillan Castle Christmas Gift Fair   
71                          Merchy Christmas   

                                             Category  \
5   {'category': ['business-and-tech', 'christmas'...   
6   {'category': ['christmas', 'family-friendly', ...   
15  {'category': ['christmas', 'family-friendly', ...   
16  {'category': ['christmas', 'family-friendly', ...   
19  {'category': ['christmas', 'family-friendly', ...   
30  {'category': ['christmas', 'family-friendly', ...   
32  {'category': ['christmas', 'family-friendly', ...   
38  {'category': ['christmas', 'family-friendly', ...   
51  {'category': ['art-and-theatre', 'film-and-lit...   
52   {'category': ['art-and-theatre', 'free-events']}   
54  {'category': ['christmas', 'family-friendly', ...   
71  {'category': ['business-and-tech', 'christmas'...   

                    Location  \
5     Stillgarden Distillery   
6          Shackleton Garden   
15  Farmleigh House & Estate   
16            Bremore Castle   
19             Dublin Castle   
30         The Mansion House   
32                    Dublin   
38      St Martin Apostolate   
51            The Substation   
52  The Douglas Hyde Gallery   
54          Ardgillan Castle   
71          The Grand Social   

                                          Description  \
5   A Christmas Market in the heart of Dublin 8 sh...   
6   Illuminate your nights with the enchanting Fes...   
15  Welcome to the Farmleigh Christmas Programme f...   
16  The perfect excuse to get into the festive spi...   
19  Come and join us this festive season to experi...   
30  The Lord Mayor of Dublin Daithí de Róiste woul...   
32  Dublin City Council is delighted to announce t...   
38  This year The Moving Crib will be opening it’s...   
51  Dublin Port Company presents Solidarity - The ...   
52  The Douglas Hyde is delighted to present a maj...   
54  Soak up the ultimate Christmas Atmosphere in A...   
71  Merchy Christmas is back! A real-life merch ma...   

                                            Image URL  \
5   https://cdn.dublin.ie/wp-content/uploads/WO-St...   
6   https://cdn.dublin.ie/wp-content/uploads/WO-Sh...   
15  https://cdn.dublin.ie/wp-content/uploads/WO-Xm...   
16  https://cdn.dublin.ie/wp-content/uploads/WO-Br...   
19  https://cdn.dublin.ie/wp-content/uploads/WO-Ch...   
30  https://cdn.dublin.ie/wp-content/uploads/WO-Li...   
32  https://cdn.dublin.ie/wp-content/uploads/WO-Du...   
38  https://cdn.dublin.ie/wp-content/uploads/WO-Th...   
51  https://cdn.dublin.ie/wp-content/uploads/WO-So...   
52  https://cdn.dublin.ie/wp-content/uploads/WO-Do...   
54  https://cdn.dublin.ie/wp-content/uploads/WO-Ar...   
71  https://cdn.dublin.ie/wp-content/uploads/WO-Th...   

                                           Event Link      sDate      eDate  
5   https://dublin.ie/whats-on/listings/stillgarde... 2023-12-09 2023-12-10  
6   https://dublin.ie/whats-on/listings/the-enchan... 2023-11-20 2023-12-10  
15  https://dublin.ie/whats-on/listings/christmas-... 2023-12-02 2023-12-17  
16  https://dublin.ie/whats-on/listings/balbriggan... 2023-11-26 2023-12-17  
19  https://dublin.ie/whats-on/listings/christmas-... 2023-12-06 2023-12-19  
30  https://dublin.ie/whats-on/listings/live-anima... 2023-12-06 2023-12-24  
32  https://dublin.ie/whats-on/listings/winter-lig... 2023-12-01 2023-12-31  
38  https://dublin.ie/whats-on/listings/the-moving... 2023-11-28 2023-01-07  
51  https://dublin.ie/whats-on/listings/solidarity... 2023-11-23

All the events above are free! So you can just search for any particular category by changing the value in the above cell.

## Filtering by location

In [ ]:
df['Location'].unique()

array(['Glass Mask Theatre', 'The Sound House', 'RDS',
       'National Botanic Gardens', 'EPIC The Irish Emigration Museum',
       'Stillgarden Distillery', 'Shackleton Garden', 'The Fumbally',
       'UCD Cinema', 'Samuel Beckett Theatre', 'Vicar Street',
       'Central Bank of Ireland', 'Richmond Barracks',
       'Goldenbridge Cemetery', 'Temple Bar Gallery + Studios',
       'Farmleigh House & Estate', 'Bremore Castle', 'Central Plaza',
       'The Casino Model Railway Museum', 'Dublin Castle',
       'Dalkey Castle & Heritage Centre', 'Smock Alley Theatre',
       'Croke Park', 'Merrion Square Park', 'Airfield Estate',
       'Dún Laoghaire', 'Malahide Castle', 'Leaves from Essex Street',
       "Bewley's Cafe Theatre", 'The Mart Gallery', 'The Mansion House',
       'The Lark Concert Hall', 'Dublin', 'The Ark', 'Millbank Theatre',
       'Dublin Zoo', 'Mill Theatre', 'St Martin Apostolate',
       'Gaiety Theatre', 'Civic Theatre', 'Bord Gáis Energy Theatre',
       'Gate Thea

Let's say I want to see all the events taking place in 'National Concert Hall' between 25 December and 02 January'24



In [ ]:
start_date = pd.to_datetime('2023-12-25')
end_date = pd.to_datetime('2024-01-02')
filtDf = df[(df['Location'] == 'National Concert Hall')& (df['sDate'] >= start_date) & (df['eDate'] <= end_date)]
filtDf


Event Name  \
83                            The Sound of Music   
86  New Year’s Eve Extravaganza: Gatsby & Beyond   

                                             Category               Location  \
83  {'category': ['art-and-theatre', 'christmas', ...  National Concert Hall   
86                 {'category': ['music-and-comedy']}  National Concert Hall   

                                          Description  \
83  After a sensational sell-out run last year, th...   
86  Get your jazz hands at the ready and ring in 2...   

                                            Image URL  \
83  https://cdn.dublin.ie/wp-content/uploads/WO-NC...   
86  https://cdn.dublin.ie/wp-content/uploads/WO-NC...   

                                           Event Link      sDate      eDate  
83  https://dublin.ie/whats-on/listings/the-sound-... 2023-12-27 2023-01-02  
86  https://dublin.ie/whats-on/listings/new-years-... 2023-12-31 2023-12-31

So, that's only what we did in this projet.

Thank You so much for being the part of this journey.

Feel free to connect me (rinshaadc@gmail.com) for any queries of concerns about this project.

In [ ]:
#Recommendation: Try with Multi Hot encoding for the Category column
